src/coordinator/tsdb/remote/compressed_codecs.go

```diff
@@ -0,0 +1,236 @@
+			Segments: replicaSegments,
+		})
+	}
+
+	start := xtime.ToNanoseconds(it.Start())
+	end := xtime.ToNanoseconds(it.End())
+
+	compressedDatapoints := &rpc.CompressedDatapoints{
+		Namespace: it.Namespace().String(),
+		StartTime: start,
+		EndTime:   end,
+		Replicas:  compressedReplicas,
+	}
+
+	tagIter := it.Tags()
+	tags := make(map[string]string)
+	for tagIter.Next() {
+		tag := tagIter.Current()
+		tags[tag.Name.String()] = tag.Value.String()
+	}
+
+	return &rpc.Series{
+		Id:         it.ID().String(),
+		Tags:       tags,
+		Compressed: compressedDatapoints,
+	}, nil
+}
+
+func segmentToBlockReader(
+	seg *rpc.Segment,
+	opts checked.BytesOptions,
+	iteratorPools encoding.IteratorPools,
+) xio.BlockReader {
+	var checkedBytesWrapperPool xpool.CheckedBytesWrapperPool
+	if iteratorPools != nil {
+		checkedBytesWrapperPool = iteratorPools.CheckedBytesWrapper()
+	}
+
+	var head, tail checked.Bytes
+	if checkedBytesWrapperPool != nil {
+		head = checkedBytesWrapperPool.Get(seg.Head)
+		tail = checkedBytesWrapperPool.Get(seg.Tail)
+	} else {
+		head = checked.NewBytes(seg.Head, opts)
+		tail = checked.NewBytes(seg.Tail, opts)
+	}
+
+	segment := ts.NewSegment(head, tail, ts.FinalizeNone)
+	segmentReader := xio.NewSegmentReader(segment)
+
+	return xio.BlockReader{
+		SegmentReader: segmentReader,
+		Start:         time.Time{},
+		BlockSize:     0,
+	}
+}
+
+var (
+	opts = checked.NewBytesOptions().SetFinalizer(
+		checked.BytesFinalizerFn(func(b checked.Bytes) {
+			b.Reset(nil)
+		}))
+
+	iterAlloc = func(r io.Reader) encoding.ReaderIterator {
+		return m3tsz.NewReaderIterator(r, m3tsz.DefaultIntOptimizationEnabled, encoding.NewOptions())
+	}
+)
+
+// SeriesIteratorFromRPC converts a grpc series to a seriesIterator
+func SeriesIteratorFromRPC(iteratorPools encoding.IteratorPools, timeSeries *rpc.Series) encoding.SeriesIterator {
+	var (
+		mutliReaderArrayPool encoding.MultiReaderIteratorArrayPool
+		multiReaderPool      encoding.MultiReaderIteratorPool
+		seriesPool           encoding.SeriesIteratorPool
+
+		idPool ident.Pool
+
+		allReplicaIterators []encoding.MultiReaderIterator
+	)
+
+	if iteratorPools != nil {
+		multiReaderPool = iteratorPools.MultiReaderIterator()
+		seriesPool = iteratorPools.SeriesIterator()
+		mutliReaderArrayPool = iteratorPools.MultiReaderIteratorArray()
+		idPool = iteratorPools.ID()
+	}
+
+	compressedValues := timeSeries.GetCompressed()
+	replicas := compressedValues.GetReplicas()
+
+	if mutliReaderArrayPool != nil {
+		allReplicaIterators = mutliReaderArrayPool.Get(len(replicas))
+	} else {
+		allReplicaIterators = make([]encoding.MultiReaderIterator, 0, len(replicas))
+	}
+
+	for _, replica := range replicas {
+		segments := replica.GetSegments()
+		blockReaders := make([][]xio.BlockReader, len(segments))
+
+		for i, segment := range segments {
+			blockReadersPerSegment := make([]xio.BlockReader, 0, len(segments))
+			mergedSegment := segment.Merged
```

Comment:
Should you use `segment.GetMerged()` here instead?

Reply:
Good catch! I was sure I forgot to fix one of these ;)